include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+     * can have for the given simulation base.
+     */
+    const uint64_t pruned_top_level_multisets{};
+    /**
+     * The maximum size of the charge space of the top cluster, given the simulation base, can be inferred by the
+     * \"stars and bars\" combinatorial idea: the solution to this analogous problem determines the maximum amount of
+     * multisets of size \f$N\f$ (where \f$N\f$ is the number of SiDBs in the layout, and therefore in the top cluster)
+     * for the given base \f$b\f$. In particular, the analogy is as follows: any such multiset can be seen as \f$N\f$
+     * stars and \f$b - 1\f$ bars separating those stars. Then, the \f$b\f$ partitions forming from these \f$b - 1\f$
+     * separators each have a respective size, adding up to \f$N\f$. Therefore each partition is associated with an
+     * amount of one of the charge states of the multiset. Now we may compute total number of possible multisets for the
+     * top cluster as the number of combinations of \f$N\f$ stars and \f$b - 1\f$ bars. Hence this is computed with the
+     * following combinatorial formula: \f$\binom{N + b - 1}{b - 1}\f$.
+     */
+    const uint64_t maximum_top_level_multisets{};
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class ground_state_space_impl
+{
+  public:
+    explicit ground_state_space_impl(const Lyt& lyt, const uint64_t max_cluster_size_for_witness_partitioning,
+                                     const sidb_simulation_parameters& phys_params) noexcept :
+            base{phys_params.base},
+            top_cluster{to_sidb_cluster(sidb_cluster_hierarchy(lyt))},
+            clustering{get_initial_clustering(top_cluster, get_local_potential_bounds(lyt, phys_params))},
+            witness_partitioning_max_cluster_size{max_cluster_size_for_witness_partitioning},
+            mu_bounds_with_error{physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 physical_constants::POP_STABILITY_ERR - phys_params.mu_plus(),
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_plus()}
+    {}
+
+    ground_state_space_result run() noexcept
+    {
+        mockturtle::stopwatch<>::duration time_counter{};
+        {
+            const mockturtle::stopwatch stop{time_counter};
+
+            while (!terminate)
+            {
+                while (!update_charge_spaces())
+                    ;
+
+                move_up_hierarchy();
+            }
+        }
+
+        const uint64_t max_multisets = maximum_top_level_multisets(top_cluster->size());
+
+        return ground_state_space_result{top_cluster, time_counter, max_multisets - top_cluster->charge_space.size(),
+                                         max_multisets};
+    }
+
+  private:
+    constexpr inline bool fail_onto_negative_charge(const double pot_bound) const noexcept
```

Comment:
I am a little bit concered about how the physics is implemented here. The reason is that if we change the defintion (e.g. error limit etc.) of the physical stability in the charge distribution, it changes for all simulators (ExGS, QuickSim, and QuickExact). But it should also change for ``CC``. Do you have an idea how to fix it?

Reply:
I would think that there is only one correct implementation, also with regards to (the signs of) error bound. So in this specific case I don't think the modularity is really neccessary, as it also conflicts with the class object `mu_bounds_with_error`  in both `ground_state_space`  and `clustercomplete` which avoids having to do many many extra additions for the error bounds (and mu+) since this is executed so many times. With a more modular implementation I don't think I would trust the compiler to optimise it all out, but maybe it does.
Curious to hear what @marcelwa thinks of this.

Reply:
A remedy could actually be a class in `sidb/simulation` that is instantiated at the start of a simulation process which creates a `mu_bounds_with_error` object just a single time. The checks could then be implemented there without sacrifice.